src/hinge.cpp

```diff
@@ -0,0 +1,96 @@
+#include <wayfire/plugin.hpp>
+#include <wayfire/signal-definitions.hpp>
+#include "wayfire/input-device.hpp"
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <vector>
+
+class wayfire_hinge : public wf::plugin_interface_t
+{
+    wf::option_wrapper_t<std::string> file_name{"hinge/filename"};
+    wf::option_wrapper_t<int> poll_freq{"hinge/poll_freq"};
+    wf::option_wrapper_t<int> flip_degree{"hinge/flip_degree"};
+    
+    std::ifstream device_file;
+    std::vector<nonstd::observer_ptr<wf::input_device_t>> inputs;
+
+    bool input_enabled = true;
+
+    bool read_device() 
+    {   
+        char buf[1024];
+        device_file.seekg(0);
+        device_file.readsome(buf, 4);
+
+        if(device_file.fail()) {
+            perror("The following error occured while trying to read the hinge sensor data");
+            return false;
+        }
+        
+        int angle;
+        sscanf(buf, "%i", &angle);
+
+        if((angle < 0) | (angle > 360)) {
+            std::cout << "Read invalid data from hinge sensor: " << angle << std::endl;
+            enable_inputs(); // So we don't get stuck with disabled inputs
+            return true;
+        }
+
+        bool new_state = angle < flip_degree;
+        if(new_state != input_enabled) {
+            if(new_state) enable_inputs();
+            else disable_inputs();
+        }
+
+        return true;
+    }
+
+    void disable_inputs() {
+        input_enabled = false;
+        for (size_t i = 0; i < inputs.size(); i++)
+        {
+            auto inp = inputs[i];
+            auto inp_type = inp->get_wlr_handle()->type;
+            if(
+            inp_type == wlr_input_device_type::WLR_INPUT_DEVICE_KEYBOARD || 
+            inp_type == wlr_input_device_type::WLR_INPUT_DEVICE_POINTER
+            ) {
+                inp->set_enabled(false);
+            }
+        }
+    }
+
+    void enable_inputs() {
+        input_enabled = true;
+        for (size_t i = 0; i < inputs.size(); i++)
+        {
+            auto inp = inputs[i];
+            inp->set_enabled();
+        }
+    }
+
+    public:
+
+        void init() override
+        {
+            device_file.open(file_name, std::ifstream::in);
+
+            auto timer = new wf::wl_timer();
```

Comment:
`wl_timer` is an RAII wrapper around the wayland timer that cleans up when the object is destroyed (i.e. when it goes out of scope). Did you mean to store this in a member variable of `wayfire_hinge` instead?

Reply:
To be honest, I don't quite understand what you mean. Does wl_timer get destroyed when init() returns?

Reply:
yes it gets destroyed when init() returns, and when it gets destroyed it releases the timer's resources.

Reply:
Hmm I don't understand why it works then. Shouldn't it delete set timeouts as well?

Reply:
I'm not too sure how the wayland timer stuff works to be honest. Seeing as this wrapper defines custom cleanup logic [that is equivalent to calling `disconnect()`](https://github.com/WayfireWM/wayfire/blob/35ff4fda9e84e87f45820dbb1871e879599d41fc/src/util.cpp#L590), my guess is that you don't actually want to trigger it here.